lib/src/Utilities.cc

```diff
@@ -428,27 +443,51 @@ std::string base64Encode(const unsigned char *bytes_to_encode,
             ((char_array_3[1] & 0x0f) << 2) + ((char_array_3[2] & 0xc0) >> 6);
         char_array_4[3] = char_array_3[2] & 0x3f;
 
-        for (int j = 0; (j < i + 1); ++j)
+        for (int j = 0; (j <= i); ++j)
             ret += charSet[char_array_4[j]];
 
-        while ((i++ < 3))
-            ret += '=';
+        if (padded)
+            while ((++i < 4))
+                ret += '=';
     }
     return ret;
 }
 
-std::vector<char> base64DecodeToVector(const std::string &encoded_string)
+std::string base64EncodeUnpadded(const unsigned char *bytes_to_encode,
+                                 unsigned int in_len,
+                                 bool url_safe)
+{
+    return base64Encode(bytes_to_encode, in_len, url_safe, false);
+}
+
+size_t base64DecodedLength(unsigned int in_len, uint8_t padding)
+{
+    return ((in_len - padding) * 3) / 4;
+}
+
+std::vector<char> base64DecodeToVector(const std::string &encoded_string,
+                                       bool *out_valid)
 {
     auto in_len = encoded_string.size();
     int i = 0;
     int in_{0};
     char char_array_4[4], char_array_3[3];
     std::vector<char> ret;
-    ret.reserve(in_len);
+    ret.reserve(base64DecodedLength(
+        in_len, in_len - (encoded_string.find_last_not_of('=') + 1)));
```

Comment:
The reserve allocates `N + 1` bytes to account for `'\0'` when calling `str.c_str()` or `str.data()`. The valid portion stays `N`.

Regarding the calculations, it is debatable, to favor memory or CPU cycles. If we omit the padding subtractions, worst case scenario we allocate 2 extra bytes. If we account for padding however, at worst it needs to do 3 `'='` checks, and perform some additions and subtractions.

Since CPU is faster than memory from what we have heard, I favored memory. Without profiling or benchmarking we can't tell.

Final decision is up to drogon maintainers.